File: Ruby/Weapon.rb
#encoding:utf-8

require_relative 'WeaponType.rb'

module Deepspace

    class Weapon

        def initialize(_name,_type,_uses)
            @name = _name
            @type = _type
            @uses = _uses
        end

        def self.newCopy(w)
            @name = w.name
            @type = w.type
            @uses = w.uses
            self
        end
        
        def type
            @type
        end

        def uses
            @uses
        end
        
        def power()
            @type.power()
        end

        def useIt()
            ret = 1.0
            if @uses > 0
                @uses = @uses - 1
                ret = self.power()
            end

            ret
        end
    end

end

File: Java/deepspace/src/main/java/deepspace/Weapon.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

/**
 *
 * @author fjaviersaezm
 */
class Weapon {
    private String name;
    private WeaponType type;
    private int uses;

    
    
    Weapon(String name, WeaponType type, int uses) {
        this.name = name;
        this.type = type;
        this.uses = uses;
    }
    
    Weapon(Weapon w) {
        this.name = w.name;
        this.type = w.type;
        this.uses = w.uses;
    }

    public WeaponType getType() {
        return type;
    }

    public int getUses() {
        return uses;
    }
    
    public float power(){
        return this.type.getPower();
    }
    
    public float useIt(){
        float ret = 1.0f;
        
        if(this.uses > 0){
            this.uses -= 1;
            ret = this.power();
        }
        
        return ret;
    }
    
    public String toString(){
        String s = String.format("This weapon has: name = %s , WeaponType = %h , uses = %d .", name,type,uses);
        return s;
    }
    
    WeaponToUI getUIVersion(){
        return new WeaponToUI(this);
    }
}
